llama-index-integrations/vector_stores/llama-index-vector-store-firestore/llama_index/vector_store/firestore/base.py

```diff
@@ -0,0 +1,218 @@
+from google.cloud.firestore_v1 import And, FieldFilter, Or
+from google.cloud.firestore_v1.base_vector_query import DistanceMeasure
+from google.cloud.firestore_v1.document import DocumentSnapshot
+from google.cloud.firestore_v1.vector import Vector
+from llama_index.core.schema import BaseNode
+from llama_index.core.vector_stores.types import (
+    BasePydanticVectorStore,
+    ExactMatchFilter,
+    FilterCondition,
+    FilterOperator,
+    MetadataFilter,
+    MetadataFilters,
+    VectorStoreQuery,
+    VectorStoreQueryResult,
+)
+from llama_index.core.vector_stores.utils import (
+    metadata_dict_to_node,
+    node_to_metadata_dict,
+)
+
+DEFAULT_BATCH_SIZE = 500
+DEFAULT_TOP_K = 10
+
+_logger = logging.getLogger(__name__)
+
+
+def _to_firestore_operator(
+    operator: FilterOperator,
+) -> str:
+    """Convert from standard operator to Firestore operator."""
+    if operator == FilterOperator.EQ:
+        return "=="
+    if operator == FilterOperator.NE:
+        return "!="
+    if operator == FilterOperator.GT:
+        return ">"
+    if operator == FilterOperator.GTE:
+        return ">="
+    if operator == FilterOperator.LT:
+        return "<"
+    if operator == FilterOperator.LTE:
+        return "<="
+    if operator == FilterOperator.IN:
+        return "in"
+    if operator == FilterOperator.NIN:
+        return "not-in"
+    if operator == FilterOperator.CONTAINS:
+        return "array-contains"
+
+    raise ValueError(f"Operator {operator} not supported in Firestore.")
+
+
+def _to_firestore_filter(
+    standard_filters: MetadataFilters,
+) -> Union[FieldFilter, Or, And, List[FieldFilter]]:
+    """Convert from standard dataclass to filter dict."""
+    firestore_filters = []
+
+    # Convert standard filters to Firestore filters
+    # Firestore filters are FieldFilter or And/Or filters
+    for f in standard_filters.filters:
+        if isinstance(f, MetadataFilter):
+            firestore_filters.append(
+                FieldFilter(f.key, _to_firestore_operator(f.operator), f.value)
+            )
+        elif isinstance(f, ExactMatchFilter):
+            firestore_filters.append(FieldFilter(f.key, "==", f.value))
+
+    if len(firestore_filters) == 1:
+        return firestore_filters[0]
+    if standard_filters.condition == FilterCondition.AND:
+        return And(filters=firestore_filters)
+    if standard_filters.condition == FilterCondition.OR:
+        return Or(filters=firestore_filters)
+
+    return firestore_filters
+
+
+class FirestoreVectorStore(BasePydanticVectorStore):
+    """Firestore Vector Store."""
+
+    stores_text: bool = True
+    flat_metadata: bool = True
+
+    collection_name: str
+    batch_size: Optional[int] = DEFAULT_BATCH_SIZE
+    embedding_key: Optional[str] = "embedding"
+    text_key: Optional[str] = "text"
+    metadata_key: Optional[str] = "metadata"
+    distance_strategy: Optional[DistanceMeasure] = DistanceMeasure.COSINE
+
+    _client: Client
+
+    def __init__(
```

Comment:
Could we support configurable project/database? An example with docstore:

https://github.com/run-llama/llama_index/blob/69d19a1e9584e426a55222058cf97427698b93b9/llama-index-integrations/storage/docstore/llama-index-storage-docstore-firestore/llama_index/storage/docstore/firestore/base.py#L29

Reply:
Is not redundant if we allow them to pass the client? 
Those constructors don't receive the client, and require additional arguments.

Reply:
A custom client can be passed through constructor, I think it would be redundant to add another initializer.

Reply:
I would argue it is more user-friendly as the user does not need to understand "firestore python SDK" for adopting the LlamaIndex Interface for using firestore.